pkg/cmd/pulumi/search.go

```diff
@@ -0,0 +1,158 @@
+// Copyright 2016-2023, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/pulumi/pulumi/pkg/v3/backend"
+	"github.com/pulumi/pulumi/pkg/v3/backend/display"
+	"github.com/pulumi/pulumi/pkg/v3/backend/httpstate"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/apitype"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/env"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/cmdutil"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/workspace"
+	"github.com/spf13/cobra"
+	auto_table "go.pennock.tech/tabular/auto"
+)
+
+type searchCmd struct {
+	orgName     string
+	queryParams []string
+
+	Stdout io.Writer // defaults to os.Stdout
+
+	// currentBackend is a reference to the top-level currentBackend function.
+	// This is used to override the default implementation for testing purposes.
+	currentBackend func(context.Context, *workspace.Project, display.Options) (backend.Backend, error)
+}
+
+func (cmd *searchCmd) Run(ctx context.Context, args []string) error {
+	interactive := cmdutil.Interactive()
+
+	if cmd.Stdout == nil {
+		cmd.Stdout = os.Stdout
+	}
+
+	if cmd.currentBackend == nil {
+		cmd.currentBackend = currentBackend
+	}
+	currentBackend := cmd.currentBackend // shadow the top-level function
+
+	opts := backend.QueryOptions{}
+	opts.Display = display.Options{
+		Color:         cmdutil.GetGlobalColorization(),
+		IsInteractive: interactive,
+		Type:          display.DisplayQuery,
+	}
+	// Try to read the current project
+	project, _, err := readProject()
+	if err != nil && !errors.Is(err, workspace.ErrProjectNotFound) {
+		return err
+	}
+
+	backend, err := currentBackend(ctx, project, opts.Display)
+	if err != nil {
+		return err
+	}
+	cloudBackend, isCloud := backend.(httpstate.Backend)
+	if !isCloud {
+		return errors.New("Pulumi AI search is only supported for the Pulumi Cloud")
+	}
+	userName, orgs, err := cloudBackend.CurrentUser()
+	if err != nil {
+		return err
+	}
+	var filterName string
+	if cmd.orgName == "" {
+		filterName = userName
+	} else {
+		filterName = cmd.orgName
+	}
```

Comment:
I generally try to avoid mutation-in-place when it can be avoided, but if this is considered more idiomatically correct, then I'm happy to make this change!

Reply:
In this case, I'm considering it less a mutation and more a "default value with special case":

```go
x := defaultValue
if condition {
  x = override
}
```

This case just happens to include validation before the override.

The other patter is indeed:

```
var x Type
if cond {
  x = a
} else {
  x = b
}
```

Which is closer to a "paragraph" that calculates the value of x. I read that closer to a block expression in some languages, e.g.

```
x = if cond { a } else { b };
```

However, for this case, we have a bit of extra validation we need to do on one of the arms of the if statement, which makes sense to do at the same time we're calculating filterName.

The following would also be okay:

```go
var filterName string
if cmd.orgName == "" {
  filterName = userName
} else {
  // TODO: sliceContains check here
  filterName = cmd.orgName
}
```

Depends on whether you read it as "a block that calculates the value of filterName" or "filterName is always userName except if an orgName was provided". I don't have a strong preference between the two.